<a href="https://colab.research.google.com/github/trevorwjames/DS-Unit-1-Sprint-2-Statistics/blob/master/DS-Unit-1-Sprint-2-Statistics%20/%20module2%20/%20Trevor_James_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [0]:

# Importing Data set 
import pandas as pd
import numpy as np

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-06-09 19:56:05--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.07s   

2020-06-09 19:56:05 (304 KB/s) - ‘student.zip’ saved [20478/20478]



In [3]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [6]:
from scipy import stats
import pandas as pd

df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [8]:
# testing different crosstabs to see is variables are dependent
# null hypothesis: Daily alchol consumption IS NOT related to test failure numbers

df['Dalc'].value_counts().sort_index()

1    276
2     75
3     26
4      9
5      9
Name: Dalc, dtype: int64

In [9]:
df['failures'].value_counts().sort_index()

0    312
1     50
2     17
3     16
Name: failures, dtype: int64

In [10]:
dalc_failuresct = pd.crosstab(df['failures'], df['Dalc'])

dalc_failuresct

Dalc        1   2   3  4  5
failures                   
0         232  50  17  6  7
1          29  12   6  2  1
2           7   8   1  1  0
3           8   5   2  0  1

In [11]:
stats.chi2_contingency(dalc_failuresct)

(22.61378441601607,
 0.031189577855521466,
 12,
 array([[218.00506329,  59.24050633,  20.53670886,   7.10886076,
           7.10886076],
        [ 34.93670886,   9.49367089,   3.29113924,   1.13924051,
           1.13924051],
        [ 11.87848101,   3.2278481 ,   1.11898734,   0.38734177,
           0.38734177],
        [ 11.17974684,   3.03797468,   1.05316456,   0.36455696,
           0.36455696]]))

Through the above test we can say with 97% confidence to rejectthe Null Hypothesis, and suggest that the consumption of alcohol is related to test failures. 

We Reject the Null Hypothesis

In [13]:
# testing if the interest in higher education is related to the amount of time students put into their studies

print(df['higher'].value_counts().sort_index())
df['studytime'].value_counts().sort_index()

no      20
yes    375
Name: higher, dtype: int64


1    105
2    198
3     65
4     27
Name: studytime, dtype: int64

In [15]:
higher_studytime = pd.crosstab(df['higher'], df['studytime'])

higher_studytime

studytime   1    2   3   4
higher                    
no         12    8   0   0
yes        93  190  65  27

In [16]:
stats.chi2_contingency(higher_studytime)

(14.187928811928812,
 0.002660195229728128,
 3,
 array([[  5.3164557 ,  10.02531646,   3.29113924,   1.36708861],
        [ 99.6835443 , 187.97468354,  61.70886076,  25.63291139]]))

It seems we have another highly correlated set of values, with 99.7% confidence that the Null Hypothesis should be rejected and suggest that the interest in higher education be related to the amount of time students place in their studies

We Reject the Null Hypothesis

In [17]:
# now to test values that would fail to reject the Null Hypothesis
# first is if there is a correlation btw the student being in a rural or urban area and their parental relationship status

address_pstatus = pd.crosstab(df['address'], df['Pstatus'])
address_pstatus

Pstatus   A    T
address         
R         7   81
U        34  273

In [18]:
stats.chi2_contingency(address_pstatus)

(0.4197410286093437,
 0.5170662942340176,
 1,
 array([[  9.13417722,  78.86582278],
        [ 31.86582278, 275.13417722]]))

From the above Chi^2 test we can see, with a pvalue of .51, the values are not related. This suggests that the two variables are independent of eachother. Urban vs Rural does not have a correlation with the Parents current relatioship status

We Fail to Reject the Null Hypothesis

In [19]:
# second test, Attendance to nursery school and Access to internet

nursery_internet = pd.crosstab(df['nursery'], df['internet'])

nursery_internet

internet  no  yes
nursery          
no        14   67
yes       52  262

In [20]:
stats.chi2_contingency(nursery_internet)

(0.0001303497288725503,
 0.9908906855790625,
 1,
 array([[ 13.53417722,  67.46582278],
        [ 52.46582278, 261.53417722]]))

With a pvalue of .99 we can suggest that there is no correlation between the attendance to nursery school and the current access to internet

we Fail to Reject the Null Hypothesis

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [22]:
nursery_internet_margins = pd.crosstab(df['nursery'], df['internet'], margins=True)

nursery_internet_margins

internet  no  yes  All
nursery               
no        14   67   81
yes       52  262  314
All       66  329  395

In [24]:
row_totals = df['nursery'].value_counts().sort_index().values 

row_totals

array([ 81, 314])

In [26]:
col_totals = df['internet']. value_counts().sort_index().values

col_totals

array([ 66, 329])

In [31]:
sample_size = df.shape[0]

sample_size

395

In [33]:
expected = np.array([[0,0],
                    [0,0]])

for i, col in enumerate(col_totals):
  for j, row in enumerate(row_totals):
    expected_value = (row*col / sample_size)
    expected[i][j] = expected_value

expected 

array([[ 13,  52],
       [ 67, 261]])

In [35]:
observed = nursery_internet.values

observed

array([[ 14,  67],
       [ 52, 262]])

In [36]:
((observed - expected)**2 / expected).sum().sum()

7.765886526694555

With a chi test value of 7.77, resulting in a pvalue of .994. We suggest that attendance to nursery school does not correlate to current access to the internet. 

we Fail to Reject the Null Hypothesis


## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?
  - An estimate as it relates to statistics is a number we create to try and infer an idea from a sample about a population. Ex, we are using a sample of students from certain schools above and could say from the chi squared tests we are performing we can infer that this could apply to all students in the population in that area. 

- What are hypothesis test? How are they useful?
  - Hypothesis tests are functions we use to test the relation, or non-relation between two observation sets of data. We have to first make hypothesises if we are going to create an experiment. They are useful for communicating findings and creating inferential ideas

- What is a "null hypothesis?"
  - A Null Hypothesis is our base statement in which we believe the status quo to be. This means it is our idea of the population that we are either failing to reject or rejecting and suggesting a differenr ideal. 

- What is a p-value and what does it represent?
  - The p-value is a probability that the null hypothesis will be true. That means the closer the p-value is to one, the closer to hypothesis is to being true. The further is is from one, while still remaining positive, means we are more likely to reject the null hypothesis and suggest an alt hypothesis.  


- What does it mean for something to be "statistically significant?"
  - Statistical signifigance is the use of the p-value to evaluate if we can effectively reject or fail to reject the null hypothesis. For the most part we use 95% confidence meaning a p-value of .05 can either swing the result to being significant or not. 

- What is a test statistic and how does it relate to a p-value?
   - The 3 tests we have been working on require a test statistic in order to find a p-value. 

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?
  - Degrees of Freedom is the sample size-1. This is used in statistical analysis to understand what possible numbers can be when filling the table. In a 1-sample ttest we need to know the entire sample size in order to calculate the degrees of freedom. In a 2 sample ttest and a chi^2 test we are comparing 2 different variables during the test. meaning n-1 == 1. This is important when it comes to testing independence so the equation knows there are only 2 variables
  

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)